### ISSUE TITLE
Waiting for Zotero database at each startup

### ISSUE THREAD

**Issue by pprw** [retorquere/zotero-better-bibtex#815]
After installing Zotero 5 and better bibtex (zotfile is also installed), I have a message at each startup saying Better Bibtex is waiting the database of Zotero.

It delays my access for no reason (I have no automatic export in place.). Is it possible to disable it (or to put it in background)?

**Comment by retorquere** [comment]
Why on earth would you assume I put it there for no reason? You think I *like* that delay?

It says "waiting for Zotero database" because it's waiting for *Zotero's* database, not mine. The auto-exports don't even need to access the Zotero DB, as they live in my own database, which is online *well* ahead of the Zotero database. I access the Zotero database (a.o.) to generate the citation keys, and the Zotero team has made it quite clear no-one is to touch the database until it has finished initializing. Having users interact with the references leads would kick off BBT activity leading to race conditions accessing the database that are really hard to recover from.

That the database now behaves this way in Zotero 5 is not my choice, nor do I have any influence over the point where Zotero gives the all-clear to extensions. Other extensions may not need the database during startup -- I do.

BBT is one of the more complex extensions out there, if not downright the most complex, modifying Zotero's behavior by patching the running program at numerous places. This is necessary because what BBT does cannot be done by the Zotero-supported extension API alone, it's a very delicate setup, and can easily break Zotero; if you post a bug report on the Zotero forums, one of the first things they'll rightfully ask is "does this also happen if you disable BBT". The wait-until-all-clear as instructed by the Zotero team is required to make the BBT/Zotero combination less fragile.

**Comment by retorquere** [comment]
Accepted. I took offense to the "unnecessary". It *is* necessary for a combination of architectural and pragmatic reasons. I dislike the idea that people would assume I had put it there without thinking things through. And I *really* dislike the idea that I should do this-or-that to avoid blame. There's an easy way for me to avoid blame, which is to close this repo, take a new email address, and go on to do things for which I am appreciated. People looking to blame me can, how shall I put this delicately, go screw themselves with a poison-tipped jackhammer. Or, you know, use the stock BibTeX built into Zotero.

**TL;DR** the "waiting" screen is there for damn good reasons. It can go away when Zotero signals me sooner that I'm allowed to touch their database. Things might get better in Zotero 5.1. Until then, it stays.

It is possible to disable BBT from the addons screen, and only re-enable it for an export, but this will require a restart of Zotero. This way, BBT will get out of your way until you need it. But if you want BBT to be ready when you need it without the enable-and-restart, it needs to be in a stable state. It's like demanding from your legislator that no money goes to ambulances you never use, but then, but then being grumpy it takes 5 days for them to order a new ambulance when you do need one.

As to the "most common case": if you feed the bib(la)tex database, you need citation keys. For citation keys, I need to access the database. For access to the database, I need to wait. This has nothing to do with auto-export. This has everything to do with how BBT works through-and-through, and how long Zotero takes to give the go-ahead. In Zotero 4, extensions would have the guarantee that the DB was ready before they were even called. In Zotero 5, there are several things that are started at the same time, and the user is given the illusion that Zotero is ready because you can click things. As soon as you actually try to do something (like adding a reference, or exporting), you'll have to wait until Zotero is *actually* ready. BBT just makes that wait visible, because I was really tired of dealing with the fallout of the race conditions that this caused. The all-clear signal is what slows down BBT startup, from a recent startup on my system:

```
Waiting for Zotero database took 11.427 s 
Loading citation key took 0.03 s 
Starting auto-export took 0.001 s 
Starting key manager took 0.846 s 
Starting serialisation cache took 0.023 s 
Loading journal abbreviator took 0 s 
Installing bundled translators took 0.685 s 
```

It's not the problem that for a second there are no citation keys. Actually it is a problem if you were to kick off an export in that time, because the export requires the cite keys to be present, but it's not *the* problem.

*The* problem is those 1.5 seconds in which Zotero has given everyone the go-ahead, and Zotero frees the interface to you, but BBT still needs to start. This has in the past lead to actual, not hypothetical, race conditions which broke the Zotero/BBT combo, for which I was also blamed. If I "buffer up" these changes to the point in time where BBT has gotten the go-ahead and finished its own initialization, that's more complexity and fragility to add to BBT, and would make it very easily the case that BBT exports are kicked off before the citekeys are all ready, which would make that export fail, and for that, of course, I would be blamed. So regardless of what I'll do, people will choose to blame me. If Zotero would give me a signal somewhere within those 12 seconds, I could *easily* startup within that time. But that is not the case.

So is it *technically* possible to get buffer-up activity within that 1.5 second time and act on it later? Yes. Is it *pragmatically* possible to do so, at the cost of quite a bit of complexity of behavior that I would then have to support for users whose systems I cannot inspect, for a single dev doing this in the wee hours alongside a family, a job, a PhD, and currently co-caring for a terminally ill family member? Occasionally trying to get some sleep?

No.

Getting BBT to work under Zotero 5 required a grounds-up rewrite which took 5 months of my life I really didn't have to spare. And all this crap is going to go another round when Zotero ports away to a new platform, soonish. I need BBT to behave in predictable ways if I am to do this. If the Zotero devs strenuously recommend waiting dor the DB, even if that means a 12 second delay before BBT can even start up its own init, that's out of my hands.

If someone is willing to volunteer to put up a patch that implements a *stable* delayed start that deals with the race conditions in a reliable way and promises to support it for the first six months in which it *will* break users' reference manager while they're under time pressure to get their thesis finished and come yelling here that I'm responsible for the disruption, then yeah, sure. But I don't see that happening, really. Of the Zotero 4 implementation, 0.102143691% came from others, the rest was me. For the Zotero 5 implementation, that number is 0.001399122212%. This is by all means a solo effort.

In the meantime, the "waiting" is necessary to keep BBTs behavior predictable to a level that can support it. If Zotero at some point manages to give the all-clear signal for DB access sooner, BBT will automatically benefit. Zotero 5.1 is going to add a separate field for cite keys and if that is actually filled always so I don't have to, a lot of problems are going to be resolved. But that is a might big "if".

**Comment by pprw** [comment]
Thanks for all these explanations. Waiting few seconds is really not a problem and you will see what's bring the 5.1 version.

I take this opportunity to thank you for the development of BBT. This very useful piece of software change the everyday life of plenty of users around the world. I was very glad to see it working when I upgraded to Zotero 5. I didn't know how I could work if it was not the case!

**Comment by retorquere** [comment]
Sorry for sounding so strongly, but getting BBT to work under Zotero 5 has *not* been a pleasant experience.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 4965](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.49.4965.issue-815.xpi) ("Merge branch 'master' into @lockless").

**Comment by retorquere** [comment]
If you want to live dangerously: https://github.com/retorquere/zotero-better-bibtex/wiki/Configuration#lockedinit
